_posts/2016-12-28-Day_Twentyone.md

```diff
@@ -0,0 +1,197 @@
+
+Since `get` is a macro, we need to understand macros before we can find out how
+that `conn` variable magically made it into our code.
+
+The word macro can mean a number of things, but in Elixir it is the mechanism
+that allows meta-programming. Meta-programming is when you write code whose
+purpose is to create other code, rather than executing. If you feel uncomfortable
+with this, you are having the right reaction: meta-programming is a very sharp
+tool that should only be used as a last resort, the main reason being how much
+harder this makes your code to understand.
+
+The compilation process in Elixir (or at least my very limited understanding of
+it!) involves generating an Abstract Syntax Tree (AST) from the source code,
+which is then fed to the compiler to generate bytecode for the virtual machine. The
+AST is simply a 'translation' of our source code into a format that is understood
+by the compiler. Interestingly, the reason why all these languages exist is
+because we cannot speak machine-language and the machine cannot speak human
+language (the source code), so the same 'content' (the program) has to be translated
+from human (the source code) all the way down to something that controls the
+electrons flowing through the processor.
+
+When we write a macro, we directly manipulate the AST. The major building block
+for macros is `quote` (which is itself a macro!); this converts any expression
+into its AST equivalent. This is a tuple with three elements; let's have a look:
+
+{% highlight elixir %}
+  iex()> quote do: sum(1,2,3)
+    # => {:sum, [], [1, 2, 3]}
+
+  iex()> quote do: variable
+    # => {:variable, [], Elixir}
+
+  iex()> quote do: (1 + 2) + 3
+    # => {:+, [context: Elixir, import: Kernel], [{:+, [context: Elixir, import: Kernel], [1, 2]}, 3]}
+{% endhighlight %}
+
+The structure has a few nuances, but in general, the first element is an atom
+identifying either the function or variable name, the second is metadata, and the
+third is the argument/s to the function or context in which a variable exists (in
+the second example above the variable `variable` exists in the global scope).
+Notice how the second `quote` above does not raise an error, even though the variable
+does not exist. This is because `quote` simply translates the given block into
+its AST representation, regardless of whether the variable we are using was
+declared or not. Finally, you can see why humans prefer using source code to
+writing ASTs: a simple expression with two additions turns into this ugly nested
+expression, which is unreasonably hard to decode.
+
+The second building block for macros is `unquote`. This is a similar mechanism to
+string interpolation, which allows us to inject a value, expression, variable,
+etc. into a string. Here is a simple example of `unquote`, which can only
+be used inside a `quote` block:
+
+{% highlight elixir %}
+  iex()> a = 2
+  iex()> quote do: 40 + a
```

Comment:
Again, may want to consider breaking this into two snippets with explanations for each.